File: sdks/node/src/plugins/copilot-io.ts
import os from 'node:os';
import path from 'node:path';
import {
  AGENT_KEY,
  CONFIG_FILE,
  type CopilotDocument,
  type CopilotSources,
  type RuntimeContract,
  createDocument,
  parseDocument,
} from './copilot-contract.js';
import { MAX_SOURCE_BYTES } from './common.js';
import {
  inspectPhysicalDirectory,
  readPhysicalFile,
  type FileSnapshot,
} from './managed-files.js';
import { managedRuntimeFilesExist } from './managed-runtime-status.js';
import {
  parseStrictJsonObject,
  parseStrictJsoncObject,
  type JsonObject,
} from './strict-json.js';

export interface CopilotPaths {
  readonly copilotHome: string;
  readonly userHooksDirectory: string;
  readonly userHookPath: string;
  readonly legacyHookPath: string;
  readonly settingsLayers: readonly SettingsLayer[];
  readonly inspectedDirectories: readonly DirectoryLocation[];
}

interface DirectoryLocation {
  readonly path: string;
  readonly label: string;
}

interface SettingsLayer {
  readonly filePath: string;
  readonly label: string;
  readonly jsonc: boolean;
}

interface ParsedSettingsLayer extends SettingsLayer {
  readonly disableAllHooks?: boolean;
  readonly snapshot?: FileSnapshot;
}

export function resolveCopilotPaths(): CopilotPaths {
  const override = process.env.COPILOT_HOME;
  const copilotHome = override ? override : path.join(os.homedir(), '.copilot');
  const project = process.cwd();
  const github = path.join(project, '.github');
  const githubCopilot = path.join(github, 'copilot');
  const githubHooks = path.join(github, 'hooks');
  const claude = path.join(project, '.claude');
  return {
    copilotHome,
    userHooksDirectory: path.join(copilotHome, 'hooks'),
    userHookPath: path.join(copilotHome, 'hooks', CONFIG_FILE),
    legacyHookPath: path.join(githubHooks, 'hooks.json'),
    settingsLayers: [
      { filePath: path.join(copilotHome, 'config.json'), label: 'legacy Copilot user config', jsonc: false },
      { filePath: path.join(copilotHome, 'settings.json'), label: 'Copilot user settings', jsonc: true },
      { filePath: path.join(claude, 'settings.json'), label: 'Claude repository settings', jsonc: true },
      { filePath: path.join(claude, 'settings.local.json'), label: 'Claude local settings', jsonc: true },
      { filePath: path.join(githubCopilot, 'settings.json'), label: 'Copilot repository settings', jsonc: true },
      { filePath: path.join(githubCopilot, 'settings.local.json'), label: 'Copilot local settings', jsonc: true },
    ],
    inspectedDirectories: [
      { path: project, label: 'Copilot working directory' },
      { path: copilotHome, label: 'COPILOT_HOME' },
      { path: path.join(copilotHome, 'hooks'), label: 'Copilot user hooks directory' },
      { path: github, label: 'GitHub configuration directory' },
      { path: githubHooks, label: 'GitHub repository hooks directory' },
      { path: githubCopilot, label: 'Copilot repository settings directory' },
      { path: claude, label: 'Claude repository settings directory' },
    ],
  };
}

async function inspectDirectories(locations: readonly DirectoryLocation[]): Promise<void> {
  for (const location of locations) {
    await inspectPhysicalDirectory(location.path, location.label);
  }
}

async function readHookDocument(
  filePath: string,
  label: string,
): Promise<CopilotDocument | undefined> {
  const snapshot = await readPhysicalFile(filePath, label, MAX_SOURCE_BYTES);
  return snapshot ? parseDocument(filePath, snapshot, label) : undefined;
}

function parseSettings(raw: string, layer: SettingsLayer): JsonObject {
  if (raw.trim().length === 0) return {};
  const label = `${layer.label} at ${layer.filePath}`;
  return layer.jsonc
    ? parseStrictJsoncObject(raw, label)
    : parseStrictJsonObject(raw, label);
}

async function readSettingsLayer(layer: SettingsLayer): Promise<ParsedSettingsLayer> {
  const snapshot = await readPhysicalFile(layer.filePath, layer.label, MAX_SOURCE_BYTES);
  if (!snapshot) return layer;
  const root = parseSettings(snapshot.contents, layer);
  if (root.disableAllHooks !== undefined && typeof root.disableAllHooks !== 'boolean') {
    throw new Error(
      `${layer.label} at ${layer.filePath} field "disableAllHooks" must be a boolean`,
    );
  }
  return {
    ...layer,
    disableAllHooks: root.disableAllHooks as boolean | undefined,
    snapshot,
  };
}

function effectiveDisabledSource(layers: readonly ParsedSettingsLayer[]): string | undefined {
  let disabledBy: string | undefined;
  for (const layer of layers) {
    if (layer.disableAllHooks === true) disabledBy = `${layer.label} at ${layer.filePath}`;
    else if (layer.disableAllHooks === false) disabledBy = undefined;
  }
  return disabledBy;
}

export async function readSources(): Promise<CopilotSources> {
  const paths = resolveCopilotPaths();
  await inspectDirectories(paths.inspectedDirectories);
  const [user, legacy] = await Promise.all([
    readHookDocument(paths.userHookPath, 'GitHub Copilot user hooks'),
    readHookDocument(paths.legacyHookPath, 'GitHub Copilot legacy project hooks'),
  ]);
  const layers = await Promise.all(paths.settingsLayers.map(readSettingsLayer));
  const userDocument = user ?? createDocument(paths.userHookPath);
  const disabledBy = userDocument.hooksDisabled
    ? `GitHub Copilot user hooks at ${paths.userHookPath}`
    : effectiveDisabledSource(layers);
  return {
    user: userDocument,
    legacy,
    disabledBy,
    settingsPreconditions: layers.map((layer) => ({
      filePath: layer.filePath,
      label: layer.label,
      snapshot: layer.snapshot,
    })),
  };
}

export function requireHooksEnabled(sources: CopilotSources): void {
  if (sources.disabledBy) {
    throw new Error(
      `GitHub Copilot hooks are disabled by ${sources.disabledBy}; set disableAllHooks to false before installation`,
    );
  }
}

export async function runtimeFilesExist(contracts: RuntimeContract[]): Promise<boolean> {
  for (const contract of contracts) {
    if (await managedRuntimeFilesExist(contract, AGENT_KEY)) return true;
  }
  return false;
}

File: sdks/go/cmd/elydora/plugins/copilot_io.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type copilotSettingsLayer struct {
	filePath string
	label    string
	jsonc    bool
}

type copilotDirectoryLocation struct {
	path  string
	label string
}

type copilotPaths struct {
	copilotHome        string
	userHooksDirectory string
	userHookPath       string
	legacyHookPath     string
	settingsLayers     []copilotSettingsLayer
	directories        []copilotDirectoryLocation
}

type parsedCopilotSettings struct {
	layer    copilotSettingsLayer
	disabled *bool
	snapshot *managedFileSnapshot
}

func resolveCopilotPaths() (*copilotPaths, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, fmt.Errorf("resolve home directory: %w", err)
	}
	override := os.Getenv("COPILOT_HOME")
	copilotHome := override
	if strings.TrimSpace(override) == "" {
		copilotHome = filepath.Join(home, ".copilot")
	} else if !filepath.IsAbs(copilotHome) {
		copilotHome, err = filepath.Abs(copilotHome)
		if err != nil {
			return nil, fmt.Errorf("resolve COPILOT_HOME at %s: %w", override, err)
		}
	}
	project, err := os.Getwd()
	if err != nil {
		return nil, fmt.Errorf("resolve working directory: %w", err)
	}
	github := filepath.Join(project, ".github")
	githubCopilot := filepath.Join(github, "copilot")
	githubHooks := filepath.Join(github, "hooks")
	claude := filepath.Join(project, ".claude")
	userHooks := filepath.Join(copilotHome, "hooks")
	return &copilotPaths{
		copilotHome:        copilotHome,
		userHooksDirectory: userHooks,
		userHookPath:       filepath.Join(userHooks, copilotConfigFile),
		legacyHookPath:     filepath.Join(githubHooks, "hooks.json"),
		settingsLayers: []copilotSettingsLayer{
			{filepath.Join(copilotHome, "config.json"), "legacy Copilot user config", false},
			{filepath.Join(copilotHome, "settings.json"), "Copilot user settings", true},
			{filepath.Join(claude, "settings.json"), "Claude repository settings", true},
			{filepath.Join(claude, "settings.local.json"), "Claude local settings", true},
			{filepath.Join(githubCopilot, "settings.json"), "Copilot repository settings", true},
			{filepath.Join(githubCopilot, "settings.local.json"), "Copilot local settings", true},
		},
		directories: []copilotDirectoryLocation{
			{project, "Copilot working directory"},
			{copilotHome, "COPILOT_HOME"},
			{userHooks, "Copilot user hooks directory"},
			{github, "GitHub configuration directory"},
			{githubHooks, "GitHub repository hooks directory"},
			{githubCopilot, "Copilot repository settings directory"},
			{claude, "Claude repository settings directory"},
		},
	}, nil
}

func inspectCopilotDirectories(locations []copilotDirectoryLocation) error {
	for _, location := range locations {
		if _, err := managedPhysicalDirectoryExists(location.path, location.label); err != nil {
			return err
		}
	}
	return nil
}

func readCopilotDocument(filePath, label string) (*copilotDocument, error) {
	snapshot, err := readManagedFile(filePath, label, maxManagedSourceBytes)
	if err != nil || snapshot == nil {
		return nil, err
	}
	return parseCopilotDocument(filePath, snapshot, label)
}

func parseCopilotSettings(
	snapshot *managedFileSnapshot,
	layer copilotSettingsLayer,
) (map[string]any, error) {
	if len(strings.TrimSpace(string(snapshot.contents))) == 0 {
		return map[string]any{}, nil
	}
	label := fmt.Sprintf("%s at %s", layer.label, layer.filePath)
	if layer.jsonc {
		return decodeJSONCObject(snapshot.contents, label, true)
	}
	return decodeStrictJSONObject(snapshot.contents, label)
}

func readCopilotSettingsLayer(
	layer copilotSettingsLayer,
) (parsedCopilotSettings, error) {
	snapshot, err := readManagedFile(
		layer.filePath,
		layer.label,
		maxManagedSourceBytes,
	)
	if err != nil || snapshot == nil {
		return parsedCopilotSettings{layer: layer}, err
	}
	root, err := parseCopilotSettings(snapshot, layer)
	if err != nil {
		return parsedCopilotSettings{}, err
	}
	result := parsedCopilotSettings{layer: layer, snapshot: snapshot}
	if value, exists := root["disableAllHooks"]; exists {
		disabled, ok := value.(bool)
		if !ok {
			return parsedCopilotSettings{}, fmt.Errorf(
				`%s at %s field "disableAllHooks" must be a boolean`,
				layer.label,
				layer.filePath,
			)
		}
		result.disabled = &disabled
	}
	return result, nil
}

func effectiveCopilotDisabledSource(layers []parsedCopilotSettings) string {
	disabledBy := ""
	for _, layer := range layers {
		if layer.disabled == nil {
			continue
		}
		if *layer.disabled {
			disabledBy = fmt.Sprintf("%s at %s", layer.layer.label, layer.layer.filePath)
		} else {
			disabledBy = ""
		}
	}
	return disabledBy
}

func readCopilotSources() (*copilotSources, *copilotPaths, error) {
	paths, err := resolveCopilotPaths()
	if err != nil {
		return nil, nil, err
	}
	if err := inspectCopilotDirectories(paths.directories); err != nil {
		return nil, nil, err
	}
	user, err := readCopilotDocument(paths.userHookPath, "GitHub Copilot user hooks")
	if err != nil {
		return nil, nil, err
	}
	legacy, err := readCopilotDocument(
		paths.legacyHookPath,
		"GitHub Copilot legacy project hooks",
	)
	if err != nil {
		return nil, nil, err
	}
	settings := make([]parsedCopilotSettings, 0, len(paths.settingsLayers))
	for _, layer := range paths.settingsLayers {
		parsed, readErr := readCopilotSettingsLayer(layer)
		if readErr != nil {
			return nil, nil, readErr
		}
		settings = append(settings, parsed)
	}
	if user == nil {
		user = createCopilotDocument(paths.userHookPath)
	}
	disabledBy := effectiveCopilotDisabledSource(settings)
	if user.hooksDisabled {
		disabledBy = fmt.Sprintf("GitHub Copilot user hooks at %s", paths.userHookPath)
	}
	preconditions := make([]copilotSourcePrecondition, 0, len(settings))
	for _, item := range settings {
		preconditions = append(preconditions, copilotSourcePrecondition{
			filePath: item.layer.filePath,
			label:    item.layer.label,
			snapshot: item.snapshot,
		})
	}
	return &copilotSources{
		user: user, legacy: legacy, disabledBy: disabledBy,
		settingsPreconditions: preconditions,
	}, paths, nil
}

func requireCopilotHooksEnabled(sources *copilotSources) error {
	if sources.disabledBy == "" {
		return nil
	}
	return fmt.Errorf(
		"GitHub Copilot hooks are disabled by %s; set disableAllHooks to false before installation",
		sources.disabledBy,
	)
}

func prepareRenderedCopilotChange(
	rendered *copilotRenderedDocument,
) (*fileChange, error) {
	if rendered == nil || !rendered.changed {
		return nil, nil
	}
	return prepareSnapshotSourceChange(
		rendered.document.filePath,
		"GitHub Copilot hook source",
		rendered.document.snapshot,
		rendered.next,
		0600,
		rendered.remove,
	)
}

func writeCopilotChanges(
	changes []*fileChange,
	label string,
	rename renameFunc,
	runtimeRoot string,
	agentDirectory string,
	preconditions []filePrecondition,
) error {
	hasChanges := false
	for _, change := range changes {
		if change != nil {
			hasChanges = true
			break
		}
	}
	if !hasChanges {
		return writeChanges(changes, label, rename, preconditions...)
	}
	if agentDirectory != "" {
		if err := EnsurePrivateDirectory(runtimeRoot); err != nil {
			return err
		}
		if err := EnsurePrivateDirectory(agentDirectory); err != nil {
			return err
		}
	}
	return writeChanges(changes, label, rename, preconditions...)
}
